File: backend/app/schemas/informatics/github_sync.py
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field


class GithubSyncSettingsUpdate(BaseModel):
    repo_url: str = Field(..., min_length=1, max_length=500)
    branch: str = Field(default="main", min_length=1, max_length=100)
    token: Optional[str] = Field(default=None, max_length=2000)
    enabled: bool = False
    interval_hours: int = Field(default=48, ge=1, le=24 * 30)
    delete_mode: str = Field(default="unpublish", pattern="^(unpublish|soft_delete)$")


class GithubSyncSettingsResponse(BaseModel):
    repo_url: str = ""
    repo_owner: str = ""
    repo_name: str = ""
    branch: str = "main"
    token_masked: str = ""
    token_configured: bool = False
    enabled: bool = False
    interval_hours: int = 48
    delete_mode: str = "unpublish"
    last_test_status: Optional[str] = None
    last_test_message: Optional[str] = None
    last_test_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


class GithubSyncTestRequest(BaseModel):
    repo_url: str = Field(..., min_length=1, max_length=500)
    branch: str = Field(default="main", min_length=1, max_length=100)
    token: Optional[str] = Field(default=None, max_length=2000)


class GithubSyncTestResponse(BaseModel):
    ok: bool
    message: str
    repo_owner: Optional[str] = None
    repo_name: Optional[str] = None
    branch: Optional[str] = None


class GithubSyncTriggerRequest(BaseModel):
    dry_run: bool = False
    force_recompile: bool = False


class GithubSyncRunItem(BaseModel):
    id: int
    trigger_type: str
    status: str
    repo_owner: str
    repo_name: str
    branch: str
    created_count: int
    updated_count: int
    deleted_count: int
    skipped_count: int
    error_summary: Optional[str] = None
    started_at: datetime
    finished_at: Optional[datetime] = None
    task_id: Optional[str] = None


class GithubSyncTaskStatusResponse(BaseModel):
    task_id: str
    state: str
    ready: bool
    successful: bool
    progress_percent: int = 0
    progress_done: int = 0
    progress_total: int = 0
    progress_phase: str = ""
    progress_current: str = ""
    created_paths: list[str] = []
    updated_paths: list[str] = []
    deleted_paths: list[str] = []
    compiled_note_ids: list[int] = []
    compile_failed: list[dict] = []
    error: Optional[str] = None

File: frontend/src/services/informatics/githubSync.ts
import { api } from "../api";

export interface GithubSyncSettings {
  repo_url: string
  repo_owner: string
  repo_name: string
  branch: string
  token_masked: string
  token_configured: boolean
  enabled: boolean
  interval_hours: number
  delete_mode: "unpublish" | "soft_delete"
  last_test_status?: string | null
  last_test_message?: string | null
  last_test_at?: string | null
  updated_at?: string | null
}

export interface GithubSyncRun {
  id: number
  trigger_type: string
  status: string
  repo_owner: string
  repo_name: string
  branch: string
  created_count: number
  updated_count: number
  deleted_count: number
  skipped_count: number
  error_summary?: string | null
  started_at: string
  finished_at?: string | null
  task_id?: string | null
}

export interface GithubSyncTaskStatus {
  task_id: string
  state: string
  ready: boolean
  successful: boolean
  progress_percent: number
  progress_done: number
  progress_total: number
  progress_phase: string
  progress_current: string
  created_paths: string[]
  updated_paths: string[]
  deleted_paths: string[]
  compiled_note_ids: number[]
  compile_failed: Array<{ path?: string; note_id?: number; error?: string }>
  error?: string | null
}

export const githubSyncApi = {
  async getSettings(): Promise<GithubSyncSettings> {
    const res = await api.get("/informatics/sync/github/settings");
    return res.data as unknown as GithubSyncSettings;
  },
  async saveSettings(payload: {
    repo_url: string
    branch: string
    token?: string
    enabled: boolean
    interval_hours: number
    delete_mode: "unpublish" | "soft_delete"
  }): Promise<GithubSyncSettings> {
    const res = await api.put("/informatics/sync/github/settings", payload);
    return res.data as unknown as GithubSyncSettings;
  },
  async testConnection(payload: { repo_url: string; branch: string; token: string }) {
    const res = await api.post("/informatics/sync/github/test-connection", payload);
    return res.data as unknown as { ok: boolean; message: string };
  },
  async trigger(dry_run = false, force_recompile = false): Promise<GithubSyncRun> {
    const res = await api.post("/informatics/sync/github/trigger", { dry_run, force_recompile });
    return res.data as unknown as GithubSyncRun;
  },
  async listRuns(limit = 10): Promise<GithubSyncRun[]> {
    const res = await api.get("/informatics/sync/github/runs", { params: { limit } });
    return res.data as unknown as GithubSyncRun[];
  },
  async getTaskStatus(taskId: string): Promise<GithubSyncTaskStatus> {
    const res = await api.get("/informatics/sync/github/task-status", { params: { task_id: taskId } });
    return res.data as unknown as GithubSyncTaskStatus;
  },
};
